### PR TITLE
[7/n] tls: integrate http outbound tls

### PR DESCRIPTION
**Pull request by jronak**
HTTP outbounds can be configured with client-provided TLS config via the `OutboundTLSConfigProvider` interface. Services can configure the outbound TLS like:
```
yarpc:
  outbounds:
    test-svc:
      http:
         url: http://127.0.0.1:5453
         tls:
           mode: enforced
           spiffe-ids:
             - spiffe-id-1
```
Spiffe-ids provided in the config above are passed to `OutboundTLSConfigProvider.ClientTLSConfig` to fetch the client TLS config for the outbound. This PR also exposes the following HTTP options:
- `OutboundTLSConfigProvider` - accepts outbound tls config provider instance
- `OutboundTLSConfiguration` - accepts outbound tls config used in the HTTP client
- `OutboundDestinationServiceName` - accepts the name of the outbound which is used in the tls dialer metrics `dest` tag

For reviews, follow this order:
- `transport/http/config.go`
- `transport/http/outbound.go`

### CODE REVIEW

File: transport/http/outbound.go [hunk 1]
```diff
@@ -121,11 +131,39 @@ func (t *Transport) NewOutbound(chooser peer.Chooser, opts ...OutboundOption) *O
...
+	o.sender = &transportSender{Client: client}
 	return o
 }
 
+func createTLSClient(o *Outbound) *http.Client {
+	transport, ok := o.transport.client.Transport.(*http.Transport)
+	if !ok {
+		// This should not happen as default yarpc http.Client uses
+		// http.Transport and it's not configurable by the user.
+		panic("failed to create http tls client, provided http.Client does not use http.Transport")
```
Comment (AllenLuUber) [inline transport/http/outbound.go:150]: do you want to print the type of `o.transport.client.Transports` in the panic? 
`panic(fmt.Sprintf("failed to create http tls client, provided HTTP client %T does not implement http.Transport"))`

File: transport/http/transport.go [hunk 2]
```diff
@@ -236,6 +238,14 @@ func ServiceName(name string) TransportOption {
 	}
 }
 
+// OutboundTLSConfigProvider returns an TransportOption that provides the
+// outbound TLS config provider.
+func OutboundTLSConfigProvider(provider yarpctls.OutboundTLSConfigProvider) TransportOption {
```
Comment (AllenLuUber) [inline transport/http/transport.go:243]: where do we use this option?

Comment (AllenLuUber) [reply transport/http/transport.go:243]: nvd. It is mention for end users to config. ignore the comment.

File: transport/http/config.go [hunk 3]
```diff
@@ -264,6 +309,15 @@ func (ts *transportSpec) buildOutbound(oc *OutboundConfig, t transport.Transport
 		}
 	}
 
+	option, err := oc.TLS.options(x.ouboundTLSConfigProvider)
+	if err != nil {
+		return nil, err
+	}
+
+	if option != nil {
```
[the remark this answers is not in the document]
Comment (jronak) [reply transport/http/config.go:317]: done

File: transport/http/outbound.go (hunk 1 above)
Comment (jronak) [reply transport/http/outbound.go:150]: done

**Review by AllenLuUber — APPROVED** [review]

**Pull request merged**